File: arrays/intervals/my_calendar_I.rb
class Node
    attr_accessor :left, :right, :s, :e

    def initialize(s, e)
        @s, @e = s, e
        @left, @right = nil, nil
    end
end

class MyCalendar
    def initialize()
        @root = nil
    end

    def book(s, e)
        !insert(s, e, @root).nil?
    end

    private
    def insert(s, e, insert_root)
        @root = Node.new(s, e) if !@root
        return @root if !insert_root

        if e <= insert_root.s
            return insert_root.left ? insert(s, e, insert_root.left) : insert_root.left = Node.new(s, e)
        elsif s >= insert_root.e
            return insert_root.right ? insert(s, e, insert_root.right) : insert_root.right = Node.new(s, e)
        else
            return nil
        end
    end
end


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar.new()
# param_1 = obj.book(start, end)

# 729. My Calendar I
# https://leetcode.com/problems/my-calendar-i/description/

require 'test/unit'
extend Test::Unit::Assertions

calendar = MyCalendar.new()
assert_equal(calendar.book(10, 20), true)
assert_equal(calendar.book(15, 25), false)
assert_equal(calendar.book(20, 30), true)
assert_equal(calendar.book(29, 31), false)
assert_equal(calendar.book(5, 9), true)
assert_equal(calendar.book(1, 6), false)

File: arrays/intervals/my_calendar_i.py
class Node:
    def __init__(self, s, e):
        self.start, self.end = s, e
        self.left, self.right = None, None


class MyCalendar:
    def __init__(self):
        self.root = None

    def book(self, start: int, end: int) -> bool:
        if not self.root:
            self.root = Node(start, end)
            return True
        return self._insert(start, end, self.root)

    def _insert(self, s, e, insert_root):
        if e <= insert_root.start:
            if insert_root.left:
                return self._insert(s, e, insert_root.left)
            else:
                insert_root.left = Node(s, e)
                return True
        elif s >= insert_root.end:
            if insert_root.right:
                return self._insert(s, e, insert_root.right)
            else:
                insert_root.right = Node(s, e)
                return True
        else:
            return False


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)

# 729. My Calendar I
# https://leetcode.com/problems/my-calendar-i/description/


calendar = MyCalendar()
assert calendar.book(10, 20) == True
assert calendar.book(15, 25) == False
assert calendar.book(20, 30) == True
assert calendar.book(29, 31) == False
assert calendar.book(5, 9) == True
assert calendar.book(1, 6) == False
